File: backend/modules/auth/auth.repository.js
import { pool } from "../../src/config/db.js";

//POST/api/auth/register 

export const findUserById = async (id) => {
    const query = `
        SELECT u.id_user, u.first_name, u.last_name, u.email, r.role_name, u.status
        FROM users u
        LEFT JOIN roles r ON u.id_role = r.id_role
        WHERE u.id_user = $1 AND u.deleted_at IS NULL
    `;
    const result = await pool.query(query, [id]);
    return result.rows[0];
};

//Find role by name 
export const findRoleByName = async (roleName) => {
    const result = await pool.query(
        `SELECT id_role FROM roles WHERE role_name = $1`, [roleName]
    )
    return result.rows[0]
}; 

//Register 
export const findUserByEmail = async (email) => {
    const query = `
    SELECT 
        u.id_user, 
        u.first_name, 
        u.last_name, 
        u.email, 
        u.password_hash,
        r.role_name
    FROM users u
    LEFT JOIN roles r ON u.id_role = r.id_role
    WHERE u.email = $1
    `; 
    const result = await pool.query(query, [email]); 
    return result.rows[0]; 
};

export const createUser = async (user) => {
    const query=`
    INSERT INTO users(id_role, first_name, last_name, email, password_hash)
    VALUES ($1, $2, $3, $4, $5)
    RETURNING id_user, first_name, last_name, email
    `; 

    const values = [
        user.id_role,
        user.first_name, 
        user.last_name,
        user.email,
        user.password
    ]; 

    const result = await pool.query(query, values); 

    return result.rows[0]; 
}; 

File: QALearnBackend/Repositories/AuthRepository.cs
using Dapper;
using QALearnAPI.Models.Entities;

namespace QALearnAPI.Repositories;

public class AuthRepository(IConfiguration cfg) : DbRepository(cfg)
{
    public async Task<UserEntity?> FindByEmailAsync(string email)
    {
        using var c = Conn();
        return await c.QueryFirstOrDefaultAsync<UserEntity>(
            @"SELECT u.id_user, u.first_name, u.last_name, u.email,
                     r.role_name, u.status, u.created_at, u.password_hash
              FROM users u LEFT JOIN roles r ON u.id_role = r.id_role
              WHERE u.email = @Email", new { Email = email });
    }

    public async Task<UserEntity?> FindByIdAsync(Guid id)
    {
        using var c = Conn();
        return await c.QueryFirstOrDefaultAsync<UserEntity>(
            @"SELECT u.id_user, u.first_name, u.last_name, u.email,
                     r.role_name, u.status, u.created_at
              FROM users u LEFT JOIN roles r ON u.id_role = r.id_role
              WHERE u.id_user = @Id AND u.deleted_at IS NULL", new { Id = id });
    }

    public async Task<Guid?> FindRoleIdByNameAsync(string roleName)
    {
        using var c = Conn();
        return await c.QueryFirstOrDefaultAsync<Guid?>(
            "SELECT id_role FROM roles WHERE role_name = @Name", new { Name = roleName });
    }

    public async Task<UserEntity> CreateUserAsync(Guid idRole, string firstName, string lastName, string email, string passwordHash)
    {
        using var c = Conn();
        return await c.QueryFirstAsync<UserEntity>(
            @"INSERT INTO users (id_role, first_name, last_name, email, password_hash)
              VALUES (@IdRole, @FirstName, @LastName, @Email, @Hash)
              RETURNING id_user, first_name, last_name, email",
            new { IdRole = idRole, FirstName = firstName, LastName = lastName, Email = email, Hash = passwordHash });
    }

    // Se usa para leer password_hash al login
    public async Task<dynamic?> FindFullByEmailAsync(string email)
    {
        using var c = Conn();
        return await c.QueryFirstOrDefaultAsync(
            @"SELECT u.id_user, u.first_name, u.last_name, u.email,
                     u.password_hash, r.role_name
              FROM users u LEFT JOIN roles r ON u.id_role = r.id_role
              WHERE u.email = @Email", new { Email = email });
    }
}
